src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -384,6 +458,55 @@ static inline struct pp_packet* ucc_tl_mlx5_mcast_buf_get_free(ucc_tl_mlx5_mcast
     return pp;
 }
 
+static inline ucc_status_t ucc_tl_mlx5_mcast_post_user_recv_buffers(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                                    ucc_tl_mlx5_mcast_coll_req_t  *req,
+                                                                    int group_id, ucc_rank_t root,
+                                                                    int coll_type,
+                                                                    int count,
+                                                                    size_t offset)
+{
+    struct ibv_recv_wr *bad_wr = NULL;
+    struct ibv_recv_wr *rwr    = comm->call_rwr;
+    struct ibv_sge     *sge    = comm->call_rsgs;
+    struct pp_packet   *pp     = NULL;
+    uint32_t            i;
+
+    for (i = 0; i < count; i++) {
+        if (NULL == (pp = ucc_tl_mlx5_mcast_buf_get_free(comm))) {
+            tl_error(comm->lib, "not enought free pp packets to cover the entire message");
+            return UCC_ERR_NO_RESOURCE;
+        }
+
+        assert(offset % comm->max_per_packet == 0);
+        pp->packet_counter = offset / comm->max_per_packet;
+        pp->qp_id          = group_id;
+        rwr[i].wr_id       = ((uint64_t) pp);
+        rwr[i].next        = &rwr[i+1];
+        sge[2*i + 1].addr  = (uint64_t)req->rptr + root * req->length + offset;
+        sge[2*i + 1].lkey  = req->recv_mr->lkey;
+        offset            += comm->max_per_packet;
+
+        if (i == count - 1) {
+            sge[2*i + 1].length = req->last_pkt_len;
+        } else {
+            sge[2*i + 1].length = comm->max_per_packet;
+        }
+    }
+
+    if (i > 0) {
```

Comment:
is it really possible that count is 0 ?

Reply:
it is a call and we must handle all scenarios that user of this function gives as an input. But I do not think it is likely. in future PR, we will have the calls into this functions and over there you can see better.

Reply:
If you think `i=0` shouldn't happen then `assert` makes it more readable and less error prone.

Reply:
Actually, having a count=0 is not a wrong input. so I would rather just keep an if statement and return successfully if the count is 0. Also, I would not say it should not get called by count=0, I rather say it is unlikely.